File: 530-minimum-absolute-difference-in-bst/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        res, stack = [], []
        curr = root
        min_v = float('inf')
        prev = None
        while (curr or len(stack)):
            while curr:
                stack.append(curr)
                curr = curr.left
            
            curr = stack.pop(-1)
            min_v = min(abs(curr.val-prev.val), min_v) if prev else min_v
            prev = curr
            curr = curr.right
        return min_v

File: 530-minimum-absolute-difference-in-bst/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int getMinimumDifference(TreeNode* root) {
        prev = nullptr;
        dfs(root);
        return res;
    }

private:
    TreeNode * prev;
    int res = INT_MAX;
    void dfs(TreeNode* root) {
        if (!root) return;
        dfs(root->left);
        if (prev) res = min(res, abs(root->val - prev->val));
        prev = root;
        dfs(root->right);
    }
};